### ISSUE TITLE
Discussion: Virtual list api

### ISSUE THREAD

**Issue by alexreardon** [atlassian/react-beautiful-dnd#1225]
We are planning on shipping virtual list support #68. However, I am not sure what the api should look like.

> 👨‍🏫 Full disclosure: I am not a virtual list expert, so please feel free to correct any inaccuracies in my understanding

### Let's get into it

One big difference from an API perspective of supporting virtual lists is that the *dragging item (`<Draggable />`)* can be removed during a drag. Currently, this is not the case. So in order to support virtual lists we need to support the ability to render a clone as the dragging item.

```js
import React from "react";
import ReactDOM from "react-dom";
import { FixedSizeList as List } from "react-window";
import { Draggable, VirtualDroppable } from "react-beautiful-dnd";

// Note: this example won't work - it is just to see what an api could look like

const Row = ({ index, style }) => (
  <Draggable draggableId={index}>
    {provided => (
      <div
        {...provided.dragHandleProps}
        {...provided.draggableProps}
        style={{
          ...style,
          ...provided.draggableProps.style
        }}
        ref={provided.innerRef}
      >
        Row {index}
      </div>
    )}
  </Draggable>
);

const Example = () => (
  <VirtualDroppable
    droppableId="droppable"
    renderDraggingItem={provided => (
      <div
        {...provided.dragHandleProps}
        {...provided.draggableProps}
        ref={provided.innerRef}
      >
        Drag Me
      </div>
    )}
  >
    {droppableProvided => (
      <List
        height={150}
        itemCount={1000}
        itemSize={35}
        width={300}
        innerRef={droppableProvided.innerRef}
      >
        {Row}
        {/* no placeholder needed */}
      </List>
    )}
  </VirtualDroppable>
);
const rootElement = document.getElementById("root");
ReactDOM.render(<Example />, rootElement);

```

### `Droppable` renders a clone?

Right now I am thinking of having a `renderDraggingItem` prop on the `Droppable` (or `VirtualDroppable`) which would be responsible for rendering the clone.

Then I think, should we always render the dragging item as a clone? I think not as things are working quite well right now, but it is something to consider. 

### Persistent?

We could get away from needing a clone if we could mark an item in a virtual list as persistent after a drag starts so the react component is never unmounted.

## `Droppable` placeholder

This would not be needed for a home list as the full list size would not collapse when an item is removed. A foreign list might need an insertion to make room... Oh dear

**Comment by theKashey** [comment]
Edge case - virtual `table`. A "row" is a `<tr><td>..</td></tr>` and has to be rendered inside a _parent_ `table` which could be create by a VirtualTable itself. 

So in short - some times you have to render a `clone` inside a virtual List, and some Lists gives you the ability to "inject" something inside, that changes how `renderDraggingItem` works - it shall be just passed down to  render props.
```js
 <VirtualDroppable
    droppableId="droppable"
    renderDraggingItem={provided => (
      <div
        {...provided.dragHandleProps}
        {...provided.draggableProps}
        ref={provided.innerRef}
      >
        Drag Me
      </div>
    )}
  >
    {({droppableProvided, VirtalDraggingItem} => (
      <List
        height={150}
        itemCount={1000}
        itemSize={35}
        width={300}
        innerRef={droppableProvided.innerRef}
        
        extraChildren={VirtalDraggingItem} // not a real API
      >
        {Row}
      </List>
    )}
  </VirtualDroppable>
);
```

That would be more or less any VirtualList library compatible, as long as removes _decision_ from DND side.

There is also another approach - _replicate item_.
1. get `ref` to the `item node`
2. create a clone of DOMTree by moved `getComputedStyles' and `getClientBoundaryRect` + `position:absolute`
3. now you can move/display this _clone_ in any place.

This is a relatively fast operation, but it would not so easy to mirror all possible changes in runtime to the clone, potentially "freezing" it. Which could be not a bad thing.

**Comment by alexreardon** [comment]
rbd could clone the item itself...
The main issue is letting the consumer control and use the `snapshot`.. 🤔

**Comment by alexreardon** [comment]
Right now a `sensor` is responsible for initiating a drag and movements during a drag. If we move to a clone strategy (which I think we will have to), then the `sensor` would only be responsible for starting a drag. Something else would need to manage the event listeners for a drag

**Comment by dimitarnestorov** [comment]
I believe you should always render a clone in a virtual list, it would save you from a couple of headaches related to items disappearing. And the user should also be able to use the snapshot, to be able add a class or some other special thing to the item that is being dragged. What do you mean by control?

**Comment by alexreardon** [comment]
The lifecyle of the snapshot and other things gets a bit complicated when moving to a clone

**Comment by alexreardon** [comment]
I am playing with this. It looks like there will be a lot of heavy lifting

![giphy-26](https://user-images.githubusercontent.com/2182637/56878181-0e5b1680-6a96-11e9-8feb-faec9c9a1a32.gif)

**Comment by alexreardon** [comment]
Whiteboarding the event listener problem with @kangweichan 

![2019-05-01 09 57 19](https://user-images.githubusercontent.com/2182637/57001912-c3104780-6bfe-11e9-9337-143f25f1c192.jpg)

**Comment by alexreardon** [comment]
```js
// new export: VirtualDroppable
<VirtualDroppable
    droppableId="droppable"
    draggingItemClone={(source: DraggableLocation, snapshot: DraggableStateSnapshot) => (
      <div>
        Dragging over {snapshot.draggingOver}
      </div>
    )}
  >
    {({provided: DroppableProvided} => (
      // 3rd party virtual list solution
      <VirtualList
        height={150}
        itemCount={1000}
        itemSize={35}
        width={300}
        innerRef={provided.innerRef}
      >
        {YourDraggableItem}
        {/* provided.placeholder is not required */}
      </List>
    )}
  </VirtualDroppable>
);
```

**Comment by tjramage** [comment]
@alexreardon — Wow. Myself and @jpsear are so excited to try this! Really awesome progress. Please let me know if / when you would like some testers, we'd be happy to help.

**Comment by alexreardon** [comment]
I have created a Github project to track the details of this

https://github.com/atlassian/react-beautiful-dnd/projects/1

**Comment by alexreardon** [comment]
@tjramage i will let you know when I get closer. I hope to do a number of beta releases

**Comment by tjramage** [comment]
@alexreardon, awesome! Thanks for making your progress so transparent with the Github project, it's really great to see 🙏

**Comment by alexreardon** [comment]
![programatic-dragging 2019-05-07 21_44_54](https://user-images.githubusercontent.com/2182637/57296860-5f5fb000-7111-11e9-87dd-8e971ed829fc.gif)

**Comment by euge** [comment]
@alexreardon Thank you thank for making progress on this. The organization, discourse, and transparency of how this project is run...its inspiring. Also the use of emojis is 🔥

**Comment by alexreardon** [comment]
`react-window`, `react-virtualised` and `react-tiny-virtual-list` all use the same technique:

- `position:absolute` on each list item with a top/left property to control it's initial poisition
- once a scroll location is reached on the list then remove / add item as needed with new left/top

**Comment by alexreardon** [comment]
I am deciding whether to try an auto detect if a list is a virtual list. However, this seems fairly brittle. I think it is a safer option to add a `mode="virtual"` prop to a `Droppable`. 

This risk in adding this prop is that people might use the flag and then stop using a virtual list and things will break for them.

**Comment by dimitarnestorov** [comment]
What thing would break if the list is not virtual? Because that means that a list without many items (no need for scroll, or only just a bit of scroll but all items are rendered) is broken upon render.

I'm putting my bets on `mode="virtual"` with a big big big ☢️⚠️☣️WARNING☣️⚠️☢️ expressing any concerns in the docs. And a detector for dev mode if the list isn't virtual.

**Comment by alexreardon** [comment]
I merged this with #68 to avoid duplication of discussion

I will now use this issue to track virtual lists

**Comment by alexreardon** [comment]
#68

**Comment by alexreardon** [comment]
`12.0.0-alpha.2` is now out which contains the first cut of virtual list support. There are no docs yet - you will need to take a look a `stories/virtual` if you want to see how to set it up

**Comment by tjramage** [comment]
@rbudnar — I also can't make proper use of virtualised lists unless it's also possible to mount/unmount droppables during a drag operation... 😭  https://github.com/atlassian/react-beautiful-dnd/issues/1317#issuecomment-504085329 

@alexreardon — do you think this would ever be technically feasible and, if so, do you plan to add in support for this at some point?

**Comment by alexreardon** [comment]
It should be technically feasible as long as a few rules are followed

**Comment by tjramage** [comment]
Ahh that's good to know, thanks @alexreardon. It would definitely make the virtual list API a million times more useful for me, and a few others, by the sounds of it.

Wish I could buy you a beer!

**Comment by UmangThapliyal** [comment]
Hello, I am fairly new to this library but it is very useful and beautiful, I have a column which has about 2k cards and I want to use this can you please point me the document where I can get some info about it. I was not able to find ```stories/virtual```

**Comment by UmangThapliyal** [comment]
Hello I tried the above provided code snippet and got the below error
![image](https://user-images.githubusercontent.com/34316267/61200353-b760f700-a6fe-11e9-839f-d5d80d88ff8c.png)

**Comment by kangweichan** [comment]
@UmangThapliyal are you using version `12.0.0-alpha.2`? The virtual list support API hasn't been officially released yet, but is currently in alpha version at the version string I mentioned above.

See more details here: https://github.com/atlassian/react-beautiful-dnd/issues/1317

**Comment by UmangThapliyal** [comment]
> @UmangThapliyal are you using version `12.0.0-alpha.2`? The virtual list support API hasn't been officially released yet, but is currently in alpha version at the version string I mentioned above.
> 
> See more details here: #1317

Yes I was trying out ```12.0.0-alpha.2``` and faced the issue

**Comment by sage911** [comment]
Will this increase performance for [@atlaskit/tree](https://atlaskit.atlassian.com/packages/core/tree)?

**Comment by alexreardon** [comment]
Latest alpha: `12.0.0-alpha.7`

**Comment by UmangThapliyal** [comment]
> Latest alpha: `12.0.0-alpha.7`

Does it support autosizer and items having different sizes

**Comment by alexreardon** [comment]
Untested, but should be fine 🤞

**Comment by UmangThapliyal** [comment]
> Untested, but should be fine 🤞

Thanks will check it out
```VirtualDroppable``` will work in 12.0.0-alpha.7?

**Comment by alexreardon** [comment]
`VirtualDroppable` is not a thing. Use this:

```jsx
<Droppable 
  droppableId="droppable"
  mode="VIRTUAL"
  whenDraggingClone={(
  provided: DraggableProvided,
  snapshot: DraggableStateSnapshot,
  source: DraggableLocation,
 ) => React.Node}
>

</Droppable
```

**Comment by UmangThapliyal** [comment]
> `VirtualDroppable` is not a thing. Use this:
> 
> ```js-jsx
> <Droppable 
>   droppableId="droppable"
>   mode="VIRTUAL"
>   whenDraggingClone={(
>   provided: DraggableProvided,
>   snapshot: DraggableStateSnapshot,
>   source: DraggableLocation,
>  ) => React.Node}
> >
> 
> </Droppable
> ```

Oh thanks I was using this only but reading the above comment I got the doubt thanks will try it.

**Comment by alexreardon** [comment]
Docs have not been written for this part yet

**Comment by UmangThapliyal** [comment]
> Docs have not been written for this part yet

Yeah, I was trying the example from ```stories/virtual```

**Comment by suanmei** [comment]
@alexreardon How long will the official version be released? I can't wait to use this great feature

**Comment by suanmei** [comment]
@alexreardon look forward to your reply

**Comment by alexreardon** [comment]
I am trudging through some issues that have been raised in testing. No firm eta at this point. I hope to have a new alpha out soon

**Comment by pathikdevani** [comment]
@alexreardon can you help to understand how exactly dnd is working with virtualization? I am using `react-dnd` & `react-window` in my project and trying to simulate the same.

I wanna know what is the thought process to make support dnd with virtualization?

Thanks in advance.

**Comment by alexreardon** [comment]
https://github.com/atlassian/react-beautiful-dnd/blob/dev/docs/patterns/virtual-lists.md

**Comment by alexreardon** [comment]
I hope to do a talk about the *how* later. It has been a large journey

**Comment by pathikdevani** [comment]
@alexreardon I have already gone through `virtual-lists` patterns doc. It's very useful as a user of `rbd`. I know _how_ is a big ask. But if you write some doc on _how_ , it will be very helpful and it will make independent logic, that will be beyod `rbd`.

Thanks, I will wait for _how_ doc.

**Comment by suanmei** [comment]
Hi, is there any API about the **Persistent**

**Comment by alexreardon** [comment]
> Hi, is there any API about the Persistent

?

**Comment by suanmei** [comment]
> Persistent?
We could get away from needing a clone if we could mark an item in a virtual list as persistent after a drag starts so the react component is never unmounted.

Just this

**Comment by alexreardon** [comment]
Using Persistent is fine. Not every major virtual list library out there supports that feature so I didn't lean on it

**Comment by suanmei** [comment]
You mentioned in this issue #216  that you will support clone mode. And I think RBD‘s core is drag and drop🤣, most dnd libraries support this ability

**Comment by alexreardon** [comment]
We'll track dragging a copy in #216

**Comment by alexreardon** [comment]
This shipped in #1487 !

**Comment by mbielecki3** [comment]
Is there any plan for release new version of https://www.npmjs.com/package/@types/react-beautiful-dnd/v/11.0.3 for this virtualized list feature?
